### PR TITLE
feat (api): push workflow

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/environment/environment_parser.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
+package environment
+
+import (
+	"io/ioutil"
+	"strings"
+	"sync"
+
+	"github.com/go-gorp/gorp"
+	"github.com/ovh/cds/engine/api/cache"
```
Comment (sguiheux) [inline engine/api/environment/environment_parser.go]: need empty line

File: engine/api/environment/environment_parser.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	//Check valid application name
+	rx := sdk.NamePatternRegex
+	if !rx.MatchString(eenv.Name) {
+		return nil, sdk.WrapError(sdk.ErrInvalidApplicationPattern, "ParseAndImport>> Application name %s do not respect pattern %s", eenv.Name, sdk.NamePattern)
+	}
+
+	//Check if app exist
+	oldEnv, errl := LoadEnvironmentByName(db, proj.Key, eenv.Name)
+	if errl != nil && sdk.ErrorIs(errl, sdk.ErrNoEnvironment) {
+		return nil, sdk.WrapError(errl, "ParseAndImport>> Unable to load application")
```
Comment (sguiheux) [inline engine/api/environment/environment_parser.go]: environment

File: engine/api/environment/environment_parser.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+// ParseAndImport parse an exportentities.Environment and insert or update the environment in database
+func ParseAndImport(db gorp.SqlExecutor, cache cache.Store, proj *sdk.Project, eenv *exportentities.Environment, force bool, decryptFunc decryptFunc, u *sdk.User) ([]sdk.Message, error) {
+	log.Info("ParseAndImport>> Import environment %s in project %s (force=%v)", eenv.Name, proj.Key, force)
+	log.Debug("ParseAndImport>> App: %+v", eenv)
+
+	//Check valid application name
+	rx := sdk.NamePatternRegex
+	if !rx.MatchString(eenv.Name) {
+		return nil, sdk.WrapError(sdk.ErrInvalidApplicationPattern, "ParseAndImport>> Application name %s do not respect pattern %s", eenv.Name, sdk.NamePattern)
```
Comment (sguiheux) [inline engine/api/environment/environment_parser.go]: environment

File: engine/api/environment/environment_parser.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/exportentities"
+	"github.com/ovh/cds/sdk/log"
+)
+
+type decryptFunc func(gorp.SqlExecutor, int64, string) (string, error)
+
+// ParseAndImport parse an exportentities.Environment and insert or update the environment in database
+func ParseAndImport(db gorp.SqlExecutor, cache cache.Store, proj *sdk.Project, eenv *exportentities.Environment, force bool, decryptFunc decryptFunc, u *sdk.User) ([]sdk.Message, error) {
+	log.Info("ParseAndImport>> Import environment %s in project %s (force=%v)", eenv.Name, proj.Key, force)
```
Comment (sguiheux) [inline engine/api/environment/environment_parser.go]: Debug

File: engine/api/pipeline/pipeline_parser.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		}
+		eg.Group = *g
+	}
+
+	allMsg := []sdk.Message{}
+	msgChan := make(chan sdk.Message, 1)
+	done := make(chan bool)
+
+	go func() {
+		for {
```
Comment (bnjjj) [inline engine/api/pipeline/pipeline_parser.go]: `for msg := range msgChan`

File: engine/api/workflow_import.go [hunk 6]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+
+		//Load project
+		proj, errp := project.Load(api.mustDB(), api.Cache, key, getUser(ctx),
+			project.LoadOptions.WithGroups,
+			project.LoadOptions.WithApplications,
+			project.LoadOptions.WithEnvironments,
+			project.LoadOptions.WithPipelines,
+		)
+		if errp != nil {
+			return sdk.WrapError(errp, "postWorkflowImportHandler>> Unable load project")
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 7]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+
+		mError := new(sdk.MultiError)
+		for {
+			hdr, err := tr.Next()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				err = sdk.NewError(sdk.ErrWrongRequest, fmt.Errorf("Unable to read tar file"))
+				return sdk.WrapError(err, "postWorkflowPullHandler>")
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 8]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+			}
+
+			log.Debug("postWorkflowPullHandler> Reading %s", hdr.Name)
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 9]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+
+			buff := new(bytes.Buffer)
+			if _, err := io.Copy(buff, tr); err != nil {
+				err = sdk.NewError(sdk.ErrWrongRequest, fmt.Errorf("Unable to read tar file"))
+				return sdk.WrapError(err, "postWorkflowPullHandler>")
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 10]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+				if err := yaml.Unmarshal(buff.Bytes(), &wrkflw); err != nil {
+					mError.Append(fmt.Errorf("Unable to load workflow %s: %v", hdr.Name, err))
+					continue
+				}
+			}
+		}
+
+		tx, err := api.mustDB().Begin()
+		if err != nil {
+			return sdk.WrapError(err, "postWorkflowPullHandler> Unable to start tx")
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 11]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+
+		allMsg := []sdk.Message{}
+		for filename, app := range apps {
+			log.Debug("postWorkflowPullHandler> Parsing %s", filename)
+			msgList, err := application.ParseAndImport(tx, api.Cache, proj, &app, true, project.DecryptWithBuiltinKey, getUser(ctx))
+			if err != nil {
+				mError.Append(err)
+			}
+			allMsg = append(allMsg, msgList...)
+			log.Debug("postWorkflowPullHandler> -- %s OK", filename)
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 12]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+		}
+
+		for filename, app := range envs {
+			log.Debug("postWorkflowPullHandler> Parsing %s", filename)
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 13]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+		}
+
+		for filename, pip := range pips {
+			log.Debug("postWorkflowPullHandler> Parsing %s", filename)
+			msgList, err := pipeline.ParseAndImport(tx, api.Cache, proj, &pip, true, getUser(ctx))
+			if err != nil {
+				mError.Append(err)
+			}
+			allMsg = append(allMsg, msgList...)
+			log.Debug("postWorkflowPullHandler> -- %s OK", filename)
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 14]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+		allMsg = append(allMsg, msgList...)
+
+		msgListString := translate(r, allMsg)
+
+		if !mError.IsEmpty() {
+			return mError
+		}
+
+		if err := project.UpdateLastModified(tx, api.Cache, getUser(ctx), proj, sdk.ProjectPipelineLastModificationType); err != nil {
+			return sdk.WrapError(err, "importPipelineHandler> Unable to update project")
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 15]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+		}
+
+		if err := tx.Commit(); err != nil {
+			return sdk.WrapError(err, "importPipelineHandler> Cannot commit transaction")
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/pipeline/pipeline_parser.go [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+	"github.com/ovh/cds/engine/api/cache"
+	"github.com/ovh/cds/engine/api/group"
+	"github.com/ovh/cds/sdk"
+)
+
+type pipeliner interface {
+	Pipeline() (*sdk.Pipeline, error)
+}
+
```
Comment (yesnault) [inline engine/api/pipeline/pipeline_parser.go]: doc

File: engine/api/workflow_import.go [hunk 17]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+		}
+
+		tx, err := api.mustDB().Begin()
+		if err != nil {
+			return sdk.WrapError(err, "postWorkflowPullHandler> Unable to start tx")
+		}
+
+		allMsg := []sdk.Message{}
+		for filename, app := range apps {
+			log.Debug("postWorkflowPullHandler> Parsing %s", filename)
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/workflow_import.go [hunk 18]
```diff
@@ -88,3 +96,142 @@ func (api *API) postWorkflowImportHandler() Handler {
...
+			msgList, err := environment.ParseAndImport(tx, api.Cache, proj, &app, true, project.DecryptWithBuiltinKey, getUser(ctx))
+			if err != nil {
+				mError.Append(err)
+			}
+			allMsg = append(allMsg, msgList...)
+			log.Debug("postWorkflowPullHandler> -- %s OK", filename)
+		}
+
+		for filename, pip := range pips {
+			log.Debug("postWorkflowPullHandler> Parsing %s", filename)
```
Comment (yesnault) [inline engine/api/workflow_import.go]: postWorkflowPushHandler

File: engine/api/pipeline/pipeline_parser.go [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+func ParseAndImport(db gorp.SqlExecutor, cache cache.Store, proj *sdk.Project, epip pipeliner, force bool, u *sdk.User) ([]sdk.Message, error) {
```
Comment (yesnault) [inline engine/api/pipeline/pipeline_parser.go]: doc

**Pull request merged**